crud/stats/module.lua

```diff
@@ -0,0 +1,234 @@
+local clock = require('clock')
+local checks = require('checks')
+local dev_checks = require('crud.common.dev_checks')
+local errors = require('errors')
+local utils = require('crud.common.utils')
+
+local StatsError = errors.new_class('StatsError', {capture_stack = false})
+
+local stats = {}
+
+local stats_registry
+local local_registry = require('crud.stats.local_registry')
+local metrics_registry = require('crud.stats.metrics_registry')
+
+local drivers = {
+    ['local'] = local_registry,
+}
+if metrics_registry.is_supported() then
+    drivers['metrics'] = metrics_registry
+end
+
+--- Check if statistics module if enabled
+--
+-- @function is_enabled
+--
+-- @treturn[1] boolean Returns true or false.
+--
+function stats.is_enabled()
+    return stats_registry ~= nil
+end
+
+--- Initializes statistics registry and enables callbacks and wrappers
+--
+-- @function enable
+--
+-- @tparam table opts
+--
+--  @tfield string driver
+--   'local' or 'metrics' (default: 'local').
+--   If 'local', stores statistics in local registry (some Lua tables)
+--   and computes latency as overall average. 'metrics' requires
+--   `metrics >= 0.9.0` installed and stores statistics in
+--   global metrics registry (integrated with exporters)
+--   and computes latency as 0.99 quantile with aging. `metrics`
+--   driver may drastically affect performance on high load.
+--
+-- @treturn[1] boolean Returns true.
+--
+function stats.enable(opts)
+    checks({ driver = '?string' })
+
+    opts = opts or {}
+    opts.driver = opts.driver or 'local'
+
+    StatsError:assert(drivers[opts.driver] ~= nil, 'Unsupported driver: %s', opts.driver)
+
+    if stats_registry == drivers[opts.driver] then
+        return true
+    end
+
+    -- Disable old driver registry, if another one was requested.
+    stats.disable()
+
+    stats_registry = drivers[opts.driver]
+    stats_registry.init()
+
+    return true
+end
+
+--- Destroys statistics registry and disable callbacks
+--
+-- @function disable
+--
+-- @treturn[1] boolean Returns true.
+--
+function stats.disable()
+    if stats.is_enabled() then
+        stats_registry.destroy()
+        stats_registry = nil
+    end
+
+    return true
+end
+
+--- Get statistics on CRUD operations
+--
+-- @function get
+--
+-- @treturn table Statistics on CRUD operations.
+--  If statistics disabled, returns {}.
+--
+function stats.get()
+    if stats_registry ~= nil then
+        return stats_registry.get()
+    end
+
+    return {}
+end
+
+--- Wrap CRUD operation call to collect statistics
+--
+-- @function wrap
+--
+-- @tparam function func
+--  Function to wrap. Errors are caught and thrown again.
+--
+-- @tparam string op_label
+--  Label of registry collectos.
+--  Use `require('crud.common.const').OP` to pick one.
+--
+-- @return First two arguments of wrapped function output.
+--
+function stats.wrap(func, op_label)
+    dev_checks('function', 'string')
+    return function(...)
```

Comment:
It will cost even for disabled metrics (varargs are slow). Could you measure how much it cost?

Reply:
In last commit "Run performance test for metrics" I added tests with performance comparison.
```
test_insert_without_stats_wrapper: requests 692815, rps 23093, errors 0
test_insert_with_stats_disabled: requests 693431, rps 23114, errors 0

test_select_with_stats_disabled: requests 238300, rps 7943, errors 0
test_select_without_stats_wrapper: requests 242640, rps 8088, errors 0
```
in CI, I've got similar results on my local machine. So it seems that it did not affects performance.

Reply:
### HP ProBook 440 G7 i7/16Gb/256SSD

|                                 | crud calls without stats wrapper | crud calls with stats disabled | crud calls with local stats | crud calls with metrics stats |
|--------------------------------|----------------------------------|--------------------------------|-----------------------------|-------------------------------|
| select pk gt (limit 10)          | 9298                             | 9461                           | 8743                        | 8353                          |
| insert                        | 33673                            | 33533                          | 32513                       | 26412                         |

### CI Runner ([run](https://github.com/tarantool/crud/runs/4536823853?check_suite_focus=true))
|                                | crud calls without stats wrapper | crud calls with stats disabled | crud calls with local stats | crud calls with metrics stats |
|--------------------------------|----------------------------------|--------------------------------|-----------------------------|-------------------------------|
| select pk gt (limit 10)         | 5760                             | 5706                           | 5330                        | 5040                          |
| insert                           | 18445                            | 18467                          | 18670                       | 16195                         |

Reply:
[PERFORMANCE REPORT.txt](https://github.com/tarantool/crud/files/7735623/PERFORMANCE.REPORT.txt)

Reply:
Discussion continues in https://github.com/tarantool/crud/pull/244#discussion_r795079619